web/template_manager.go

```diff
@@ -0,0 +1,71 @@
+package web
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"io/fs"
+	"path/filepath"
+)
+
+const (
+	base       = "templates/base.html.tmpl"
+	blocksHtml = "templates/blocks/*.html.tmpl"
+)
+
+// the main motivation of this code is to parse all templates contained in templates directory,
+// respecting the "base" template which is the basis, where other custom templates will inherit and override content
+// we also extend the template and build the base template with blocks templated (contained in /templates/blocks directory)
+
+// AddAllTemplateFromFs return a map of parsed templated from filesystem
+// the map of parsed html templates  is consumed by http handler, accessed via key
+func AddAllTemplatesFromFS(templatesFS fs.FS, templates ...string) map[string]template.Template {
+	templs := make(map[string]template.Template)
+
+	for _, pattern := range templates {
+		files, err := fs.Glob(templatesFS, pattern)
+		if err != nil {
+			// we exceptionally hard panic in case of glob errors, these should never happen.
+			panic(err)
+		}
+		for _, file := range files {
+			if file == base {
+				continue
+			}
+			addFileFromFS(templatesFS, file, templs)
+		}
+	}
+	return templs
+}
+
+// addFileFromFS parses the base template with the user
+func addFileFromFS(templatesFS fs.FS, file string, templs map[string]template.Template) {
+	var tmpl *template.Template
+	// use the base template first
+	name := filepath.Base(file)
+	tmpl = template.New(filepath.Base(base))
+
+	// we "extend" the templates by adding custom functions
+	tmpl = tmpl.Funcs(template.FuncMap{
+		"escapedTemplate": func(name string, data interface{}) string {
+			var out bytes.Buffer
+			_ = tmpl.ExecuteTemplate(&out, name, data)
+			return out.String()
+		},
+	})
+	// parse all templates
+	patterns := append([]string{base, file}, []string{blocksHtml}...)
+	tmpl = template.Must(tmpl.ParseFS(templatesFS, patterns...))
+
+	// add template to template map, consumed by handlers
```

Comment:
Why have you added the `addTemplate` code here?
Now the method is bigger, I personally preferred having it as other method

Reply:
previously we were having 2 issues:

1) we were using `methods`  but we didn't have any state to manipulate ( https://flaviocopes.com/golang-methods-or-functions/)

2) I merged the 2 methods because  of 2 reasons:

- is better to have deeper functions ( and bigger)  rather then shorter one.
-  conceptually having 2 methods initially, is where supposed to do 2 different operations but the function we merged where just helper function. In our code, is better to don't create abstraction when we don't need them.

Reply:
> 1. we were using `methods`  but we didn't have any state to manipulate ( https://flaviocopes.com/golang-methods-or-functions/)

That's not true, previously we were building a struct with an interface associated with it, which abstracted both the behaviour and the data, now you're just returning data and passing that around everywhere it is needed.
I don't find any of the two approaches superior to the other, but we didn't need to pass all the templates to all the HTTP handlers before, so IMHO the new interface to the templates is less convenient than the previous one, considering we also have to check for slice keys every time. You will have to rebuild yet another abstraction that we already had in place.

> 2\. I merged the 2 methods because  of 2 reasons:
> 
> * is better to have deeper functions ( and bigger)  rather then shorter one.
> * conceptually having 2 methods initially, is where supposed to do 2 different operations but the function we merged where just helper function. In our code, is better to don't create abstraction when we don't need them.

That's actually totally the opposite of the best practices I would follow, which is to have little functions that are easy to read and understand because they only do one thing and very well, pretty much like the entire UNIX philosophy. You are a proponent of minimalism yourself, so I find this change very contradictory. 😛 

At the end of the day I share Xabi's view here: in-lining those two functions makes the code harder to read and understand, so I would keep them separated.

Reply:
Small functions are not easier to read. Especially here, we are creating a function for the sake of building a `container` for code which is used only 1once. It should be not even a function. A function is not a place holder for code,
 but in reality the function is just used once. There is not really anything with the unix principles. 

please read this: 

https://copyconstruct.medium.com/small-functions-considered-harmful-91035d316c29

Reply:
https://user-images.githubusercontent.com/10886597/111984614-e18dcd80-8b0b-11eb-9773-cf92560279ac.png

Reply:
@MalloZup I have stumbled upon that article in the past and I must say I simply don't agree with most of it.
The only good take away I can find is that we should not be dogmatic, which me and Xabi hardly are: those additional functions were extracted empirically, as a refactoring of a function that became too long and thus harder to explain, which I actually had to do during a live code review session between me and @arbulu89.

You are basically rewriting the code that I already wrote and then refactored, back to its non re-factored form, which I frankly find difficult to consider as an improvement.

Since you have two colleagues expressing a personal preference here, could you please be so kind to make life easier for us all?

Reply:
in this context:

I don't think that 2 colleagues vs 1 one opinion, is literally a strong point to make technical choice.
However for the simplicity/realistic mind, I will bring it back the code as the other reviewers, I think for the sake of the present PR this refactoring is `secondary` but as intellectual and open learning discussion is still valid.

Although I don't agree with your points and I find unfortunate that you miss the occasion to think/discuss about it.

example for a big function from terraform: https://github.com/hashicorp/terraform/blob/main/backend/local/backend_local.go#L32-L139

@stefanotorresi why you don't agree with the article?  There are also other books  who plaid in favor of such argument. In my experience, there are many occasion where small function are suggested as good, but in many codebases they are just a PITA to maintain and understand in term of complexity. ( that is also called spaghetti code).

Refactoring is given a shape or structure to a code. We need to consider that also when a given shape is modeled on a codebase, this has a several disavantages.

So saying `refactoring` as an improvement, is wrong because refactoring can harm the cognitive perception of the code as a "pure" bigger function is much easier to understand.

Also in this context, we are building a function for code which is used only 1 in its lifetime, isn't really a function.

Reply:
as discussed, I will bring back to the original refactoring so we can discuss other things

Reply:
Let's agree to disagree, Dario.